File: src/arrays/search_rotated_array.py
from typing import List


class Solution:

    def search(self, nums: List[int], target: int) -> int:
        leftIndex = 0
        rightIndex = len(nums) - 1

        while leftIndex <= rightIndex:
            middleIndex = (leftIndex + rightIndex) // 2
            if target == nums[middleIndex]:
                return middleIndex
            # If we are in the left subarray
            if nums[leftIndex] < nums[middleIndex]:
                # Logic to go to the subarray right or left to the middleIndex
                if target > nums[middleIndex] or target < nums[leftIndex]:
                    leftIndex = middleIndex + 1
                else:
                    rightIndex = middleIndex - 1
            # If we are in the right subarray
            else:
                if target < nums[middleIndex] or target > nums[rightIndex]:
                    rightIndex = middleIndex - 1
                else:
                    leftIndex = middleIndex + 1
        return -1

File: java/src/main/java/com/blind75/arrays/SearchRotatedArray.java
package com.blind75.arrays;

public class SearchRotatedArray {

    public int search(int[] array, int target) {
        var left = 0;
        var right = array.length - 1;

        while (left <= right) {
            var middle = (left + right) / 2;
            if (array[middle] == target) {
                return middle;
            }

            // If middle is in the left subarray
            if (array[left] <= array[middle]) {
                if (target > array[middle] || target < array[left]) {
                    left = middle + 1;
                } else {
                    right = middle - 1;
                }
            }
            // If middle is in the right subarray
            else {
                if (target < array[middle] || target > array[right]) {
                    right = middle - 1;
                }
                else {
                    left = middle + 1;
                }
            }

        }
        return -1;
    }
}
